src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -336,6 +347,11 @@ private struct Entry {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
             }
 
+            // save field accesses
+            int[] buckets = this.buckets;
+            Entry[] entries = this.entries;
+            IEqualityComparer<TKey> comparer = this.comparer;
```

Comment:
These lines must come after the Initialize(0) call below. As is now the code will crash when you try to insert something.